### ISSUE TITLE
Optimize template-based layer classes

### ISSUE THREAD

**Issue by bvanessen** [LLNL/lbann#83]
Optimize the layer classes to properly use their template type to specialize the data layout.  This needs to be done after the layer classes are refactored.

**Comment by ndryden** [comment]
We do some copying in places (e.g. softmax/categorial cross entropy, id activations) that used to be done in-place; we should try to avoid that.

**Comment by ndryden** [comment]
I think we can consider this done now that we've lifted the last vestiges of the non-template-based layer stuff out of `fully_connected::fp/bp_compute`.

There is some opportunity to use template-based typedefs to avoid having to explicitly specialize the matrix allocation in some layers, but that would be a more extensive rejiggering in my opinion.
